### ISSUE TITLE
Return to LSTM & GRU Layers + ConvTranspose 1D/2D

### ISSUE THREAD

**Issue by rep-stosw** [Dobiasd/frugally-deep#435]
Hi Tobias!

I noticed that since frugally-deep-0.15.31 recurrent layers are not supported. Why? New syntax of Tensorflow/Keras or another reasons?
Recurrent layers are important for me, because speech/DSP signal processing (time data) need it.

The second question about ConvTranspose 1D/2D layers. I read in FAQ what  ConvTranspose is not good thing, but there are more targets, there these layers are neccesary.

Please read this: https://discuss.pytorch.org/t/upsample-conv2d-vs-convtranspose2d/138081
In this discussion people says that Conv + UpScale != ConvTranspose. We lost 3/4 training parameters with Conv+UpScale.

I confirm it. I am developing AAE model and decoder scheme without ConvTranspose gives non-learning result: reconstruct loss not reduced in Conv+UpScale case.

Please, add Conv Transpose for 1D & 2D.

This code gives excellent loss reduction during training:

`decoder  = keras.Sequential(
    [
        keras.Input(shape=(k, k, f)),
        layers.Flatten(),
        layers.Dense(7 * 7 * 64, activation="relu"),
        layers.Reshape((7, 7, 64)),
        layers.Conv2DTranspose(64, 3, activation="relu", strides=2, padding="same"),
        layers.Conv2DTranspose(32, 3, activation="relu", strides=2, padding="same"),
        layers.Conv2DTranspose(1, 3, activation="sigmoid", strides=1, padding="same"),
    ],
    name="decoder",
)
`

This alternative has not loss ruduction:

`decoder  = keras.Sequential(
    [
        keras.Input(shape=(k, k, f)),
        layers.Flatten(),
        layers.Dense(7 * 7 * 64, activation="relu"),
        layers.Reshape((7, 7, 64)),
        layers.Conv2D(64, 3, activation="relu", strides=1, padding="same"),
        layers.UpSampling2D((2)),
        layers.Conv2D(32, 3, activation="relu", strides=1, padding="same"),
        layers.UpSampling2D((2)),
        layers.Conv2D(1, 3, activation="sigmoid", strides=1, padding="same"),
    ],
    name="decoder",
)
`

**Comment by Dobiasd** [comment]
Hi rep-stosw,

yeah, after [the discussion in this issue](https://github.com/Dobiasd/frugally-deep/issues/417#issuecomment-2057040382) (reasons mentioned there), I [dropped support for recurrent/stateful layers](https://github.com/Dobiasd/frugally-deep/pull/416).

Re-adding them, at least in the short/mid-term, feels out-of-scope for me to do personally.

ConvTranspose seems more doable. Could you extend the following

```python3
import keras
from keras import layers

good_model = keras.Sequential([
    keras.Input(shape=(k, k, f)),
    layers.Flatten(),
    layers.Dense(7 * 7 * 64, activation="relu"),
    layers.Reshape((7, 7, 64)),
    layers.Conv2DTranspose(64, 3, activation="relu", strides=2, padding="same"),
    layers.Conv2DTranspose(32, 3, activation="relu", strides=2, padding="same"),
    layers.Conv2DTranspose(1, 3, activation="sigmoid", strides=1, padding="same"),
], name="decoder", )

bad_model = keras.Sequential([
    keras.Input(shape=(k, k, f)),
    layers.Flatten(),
    layers.Dense(7 * 7 * 64, activation="relu"),
    layers.Reshape((7, 7, 64)),
    layers.Conv2D(64, 3, activation="relu", strides=1, padding="same"),
    layers.UpSampling2D((2)),
    layers.Conv2D(32, 3, activation="relu", strides=1, padding="same"),
    layers.UpSampling2D((2)),
    layers.Conv2D(1, 3, activation="sigmoid", strides=1, padding="same"),
], name="decoder", )
```

to a runnable minimal example, demonstrating the problem with the second version?

**Comment by Dobiasd** [comment]
Thanks for the very nice example!

I can confirm your results:

`decoder = good_model`:

```
Epoch 10/10
2142/2142 ━━━━━━━━━━━━━━━━━━━━ 22s 10ms/step - discEnc_loss: 2.2180e-08 - disc_loss: 5.2103e-08 - reconstruction_loss: 0.0166 - total_loss: 0.0166
```

`decoder = bad_model`:

```
Epoch 10/10
2142/2142 ━━━━━━━━━━━━━━━━━━━━ 20s 9ms/step - discEnc_loss: 4.7019e-11 - disc_loss: 1.6843e-08 - reconstruction_loss: 0.1307 - total_loss: 0.1307
```

I'll try to understand it more in-depth, and I will get back to you here.

**Comment by Dobiasd** [comment]
So far, I have not yet found a way to achieve the small loss in your example using `Conv2D`+`UpSampling2D` instead of `Conv2DTranspose`.

Thus, I think it might be a valid use case to implement `Conv2DTranspose` in frugally-deep. Are you interested in giving it a try?

**Comment by rep-stosw** [comment]
> Thus, I think it might be a valid use case to implement `Conv2DTranspose` in frugally-deep. Are you interested in giving it a try?

Yes, I'm ready to test Conv1DTranspose & Conv2DTranspose in my projects & another examples.
Also need Conv1DTranspose for Generative Decoders for speech/audio processing.

Please, make these implementations ConvTranspose 1D/2D for frugally-deep!
Thanks.

**Comment by Dobiasd** [comment]
Ah, I meant if you would like to contribute to frugally-deep by implementing these layers. 😁

But if not, I'll see what I can do. How urgent is it?

**Comment by rep-stosw** [comment]
> Ah, I meant if you would like to contribute to frugally-deep by implementing these layers. 😁
> 
> But if not, I'll see what I can do. How urgent is it?

To tell the truth, I only know C well, I know C++ pretty mediocrely, and I just recently started to figure out Python.

So I won't be able to add ConvTranspose to frugally-deep on my own quickly.

I'd be grateful if you could add these layers to frugally-deep. Preferably ASAP.

**Comment by Dobiasd** [comment]
Ok, no worries. I'll work on it and will get back to you. 🧑‍🔧

**Comment by Dobiasd** [comment]
Progress update: Locally, I have `Conv1DTranspose` and `Conv2DTranspose` working correctly for different kernel sizes, padding, and input depths. I'm still struggling with the implementation details for `strides != 1`, but I'm on it.

**Comment by Dobiasd** [comment]
Ok, I figured it out. :relieved:

`Conv1DTranspose` and `Conv2DTranspose` are not supported in frugally-deep. :tada:

Please check out the latest version ([v0.17.0](https://github.com/Dobiasd/frugally-deep/releases/tag/v0.17.0)) and let me know if it works for you.

**Comment by rep-stosw** [comment]
> Please check out the latest version ([v0.17.0](https://github.com/Dobiasd/frugally-deep/releases/tag/v0.17.0)) and let me know if it works for you.

OK! A'll check ASPA and write result of testing.

**Comment by rep-stosw** [comment]
Just tested Conv1DTranspose. All OK.

Few later I'll check Conv2DTranspose.
